### DISCUSSION TITLE
Oneshot perfromance

### DISCUSSION THREAD

**Discussion by tkaitchuck** [tokio-rs/tokio#3410] [Q&A]
I have a rather large and [somewhat complex application](https://github.com/pravega/pravega-client-rust/) that is getting completely bottlenecked on the performance of `oneshot`s.

High level overview of what's happening:
1. oneshot is created and returned to caller.
2. Onshot is passed through an mpsc to a previously spawned task.
3. unrelated stuff happens. (involves another handoff with an mpsc)
4. Same task from step 2 completes the oneshot.

The caller calls the function to create the oneshot and then awaits it and then awaits the oneshot over and over in a loop.

Flame graph is attached. (unfortunately git doesn't allow svgs so it's just a PNG zoomed in)
![Screenshot at 2021-01-11 17-05-06](https://user-images.githubusercontent.com/4756752/104256137-78188100-542f-11eb-8064-1ef927f9103b.png)

What's odd about that is that the same code path actually uses two `mpsc`s and those barely register on the flamegraph at all. I tried just replacing the oneshots with mpscs but actually was slower. 

Is there some way to speed up this sort of "creating, using, and dropping" oneshots in a loop? Maybe a way to avoid the `wake_by_ref` call which for whatever reason don't seem to be happening with `mpsc`s when they are used for many messages?

**Comment by Darksonn** [comment]
Did you try out other channel implementations? Some options are:

 * futures crate
 * async-channel crate
 * flume crate

**Reply by tkaitchuck** [reply]
I tried the futures crate. The perf was similar. I'll look into the other two.
